SpecialWave added

Cataract92 opened the pull request:
Refactored Specials
Added 5. SpecialWave

Cataract92 committed:
SpecialWave added

Refactored Specials
Added 5. SpecialWave

Frotty commented on wurst/Creeps/UnitSpawnSystem.wurst:
```diff
@@ -21,12 +26,11 @@ int level = 1
 	
 		
 function onSpawnTime()	
-	if level mod 7 == 0 and specialWaves.getSize() > 0
```
please re-add these checks to avoid nullpointer errors

Frotty commented on wurst/Creeps/specials/Special.wurst:
```diff
@@ -0,0 +1,81 @@
+import Creep
+import ObjEditingNatives
+
+//COMPILETIME
```
Use the Human Readable classes please (UnitDefinition/AbilityDefinition)

Frotty commented on wurst/_abstract/buff/Buff.wurst:
```diff
@@ -63,35 +63,44 @@ init
 						for bff in attackerData.getBuffs()
 							if not bff.done
 								bff.attackModifier()
-					modifyDamage(0)
```
Remove this and give the SPecial a Buff extended NormalShieldBuff instead. use defenseModifier as onDamage

Frotty commented on wurst/_abstract/entity/Entity.wurst:
```diff
@@ -130,6 +131,8 @@ abstract public class UnitEntity extends Entity
 	function onDeath(unit killer)
 		done = true
 			
+	function onDamage(unit attacker)
```
same for these modifications

Frotty commented on wurst/Creeps/specials/Special.wurst:
```diff
@@ -0,0 +1,81 @@
+//
+
+public class Special extends Creep
+	effect glow
```
make class abstract

Frotty commented on wurst/Creeps/specials/SpecialFive.wurst:
```diff
@@ -0,0 +1,42 @@
+import PresetBuffs
+import Special
+
+class SlowDebuff extends NormalBuff
```
extend the NormalShieldBuff and then the defense modifier for onDamage

Frotty commented on wurst/Creeps/specials/SpecialFive.wurst:
```diff
@@ -0,0 +1,42 @@
+		super(victim, dur, abilId, true)
+
+	override function defenseModifier()
+		super.defenseModifier()
```
These super calls are empty, please remove

Frotty commented on wurst/Creeps/specials/SpecialFive.wurst:
```diff
@@ -0,0 +1,42 @@
+
+public class SpecialFive extends Special
+	
+	static constant real SPAWNTIME = 20
```
you can leave out types on constants

Frotty commented on wurst/Creeps/specials/SpecialTwo.wurst:
```diff
@@ -0,0 +1,28 @@
+
+	override function onDeath(unit killer)
+		var ppos = getPos()
+		if actor.getTypeId() == 'u00B'
```
Put ids into constants please

Frotty commented on wurst/_abstract/entity/Entity.wurst:
```diff
@@ -75,6 +75,7 @@ public abstract class Entity
 		
 		let startZVelocity = ((-gravity.z * t) / 2 - pos.z/t + e/t)
 		this.setVel( vec3( Cos(tangle.radians()) * speed, Sin(tangle.radians()) * speed, startZVelocity) )
+		real i = ANIMATION_PERIOD + 0
```
??

Cataract92 replied:
This was just a test I forgot to remove completly

Frotty commented on wurst/Creeps/specials/SpecialFive.wurst:
```diff
@@ -0,0 +1,42 @@
+		if spawntimer >= SPAWNTIME
+			spawntimer = 0
+			for i = 1 to SPAWNAMOUNT
+				new Spawn(actor.getPos3(),SPAWNID,0,middle)
```
Spawning multiple units at once looks ugly and might block the caster, instead let them come after each other. Maybe keep the Special in some channel/cast animation while it spawns the minions.
Maybe also a sound.

Frotty commented on wurst/Creeps/specials/SpecialFive.wurst:
```diff
@@ -0,0 +1,42 @@
+
+class Spawn extends Special
+	construct(vec3 pos, int typId, int bounty, vec2 target)
+		super(pos, typId, bounty, target)
```
Make the appearance more fancy, play spawn animation, some special effect

Cataract92 committed:
Reworked DmgDetection + Spawn Animations (Special 5)

Removed onDamage(unit attacker) and added BoundBuff for DmgDetection
added Spawn Animation for Special 5
rewrote compiletime functions
some refactoring

Frotty commented on wurst/Creeps/UnitSpawnSystem.wurst:
```diff
@@ -24,9 +29,8 @@ function onSpawnTime()
-		if normalWaves.getSize() > 0
-			normalWaves.dequeue().spawn()
-		if level mod 4 == 0 and normalWaves.getSize() > 0
+		normalWaves.dequeue().spawn()
```
check if it has members before dequeuing to avoid errors at the end of the game.

Cataract92 committed:
Implemented IdGenerator. Added checks while spawn

Dynamic generation of IDs generated in compiletime
added checks while spawning waves

Frotty commented:
nice thx
